stdlib/public/TensorFlow/TensorHandle.swift

```diff
@@ -16,33 +16,49 @@
 
 import CTensorFlow
 
-/// `_AnyTensorHandle` is the scalar-agnostic base type for `TensorHandle`, used
-/// specifically for low-level, type-erased passings of Swift-level tensor
-/// handles in the compiler.
-@_fixed_layout // required because the compiler accesses _cTensorHandle directly.
-public class _AnyTensorHandle {
+/// This protocol abstracts the underlying representation of a tensor. Any type
+/// that conforms to this protocol can be used as a `TensorHandle` in the
+/// `TensorFlow` library, as it much provide a way to convert the underlying tensor
+/// handle into a `ConcreteTensorHandle`, which wraps a `TFE_TensorHandle *`
+public protocol _AnyTensorHandle {
+  var tfeTensorHandle: TFETensorHandle { get }
+}
+
+extension _AnyTensorHandle {
   /// The underlying `TFE_TensorHandle *`.
-  ///
-  /// - Note: The compiler knows that `_AnyTensorHandle` has a single stored
-  /// property, and assumes that this is it. Changing the design of
-  /// `TensorHandle` will require tweaking the compiler.
+  public var _cTensorHandle: CTensorHandle {
+    return tfeTensorHandle._cTensorHandle
+  }
+}
+
+/// Class wrapping a C pointer to a TensorHandle.  This class owns the
+/// TensorHandle and is responsible for destroying it.
+public class TFETensorHandle : _AnyTensorHandle {
   public let _cTensorHandle: CTensorHandle
 
-  /// Private initializer from a `CTensorHandle`. Should only be called from
-  /// `TensorHandle<Scalar>.init`.
-  fileprivate init(base: CTensorHandle) {
+  public var tfeTensorHandle: TFETensorHandle { return self }
+
+  public init(_owning base: CTensorHandle) {
     self._cTensorHandle = base
   }
+
+  deinit {
+    debugLog("De-initializing TensorHandle.")
+    TFE_DeleteTensorHandle(_cTensorHandle)
+    debugLog("Returning from deinit of TensorHandle.")
+  }
 }
 
-/// `TensorHandle` is the type used by ops. It includes a `Scalar` type, which 
-/// compiler internals use to determine the datatypes of parameters when they 
-/// are extracted into a tensor program.
-@_fixed_layout // required because the compiler accesses _cTensorHandle directly.
-public final class TensorHandle<Scalar> : _AnyTensorHandle
-  where Scalar : _TensorFlowDataTypeCompatible {
+/// `TensorHandle` is the type used by ops. It includes a `Scalar` type, which
+/// compiler internals can use to determine the datatypes of parameters when
+/// they are extracted into a tensor program.
+public final class TensorHandle<Scalar> where Scalar : _TensorFlowDataTypeCompatible {
```

Comment:
Since `TensorHandle` no longer possesses a `deinit` and that it wraps a reference pointer, it need no reference semantics or heap allocation. Therefore, for efficiency, `TensorHandle<T>` should be turned into a `struct` as well as `ResourceHandle` and `VariantHandle`.

Reply:
Changing them to struct causes the compiler to crash: https://bugs.swift.org/browse/TF-527. 

Let me submit this PR without the struct change and I will investigate this later?

Reply:
Sounds good! Looks like an AD bug.